chromatic_fitting.py

```diff
@@ -78,127 +86,277 @@ def summarize_parameters(self):
+            data_copy.wavelike[k] = [data_copy.wavelike[k][i]]
+        return data_copy
+
+    def get_data(self, i=0):
         """
         Extract the data to use for the optimization depending on the method chosen
         """
         if hasattr(self, 'optimization'):
             if self.optimization == "white_light":
                 self.white_light_curve()
                 return self.white_light
+            if self.optimization == 'separate':
+                return self.separate_wavelengths(i)
         return self.data
 
     def setup_likelihood(self):
         """
         Connect the light curve model to the actual data it aims to explain.
         """
-        data = self.get_data()
+        # data = self.get_data()
 
-        with self.model:
-            for i, w in enumerate(data.wavelength):
-                k = f"wavelength_{i}"
-                # mu = Deterministic(f'{k}_mu',self.every_light_curve[k])
-                pm.Normal(
-                    f"{k}_data",
-                    mu=self.every_light_curve[k],
-                    sd=data.uncertainty[i, :],
-                    observed=data.flux[i, :],
-                )
+        if self.optimization == "separate":
+            models = self.pymc3_model
+            datas = [self.get_data(i) for i in range(self.data.nwave)]
+        else:
+            models = [self.pymc3_model]
+            datas = [self.get_data()]
+
+        for j, (mod, data) in enumerate(zip(models, datas)):
+            with mod:
+                for i, w in enumerate(data.wavelength):
+                    k = f"wavelength_{j+i}"
+                    # mu = Deterministic(f'{k}_mu',self.every_light_curve[k])
+                    pm.Normal(
+                        f"{k}_data",
+                        mu=self.every_light_curve[k],
+                        sd=data.uncertainty[i, :],
+                        observed=data.flux[i, :],
+                    )
 
     def sample_prior(self, ndraws=3):
         """
         Draw samples from the prior distribution.
         :parameter n
         Number of priors to sample
         """
-        with self.model:
-            return sample_prior_predictive(ndraws)
+        try:
+            with self.pymc3_model:
+                return sample_prior_predictive(ndraws)
+        except:
+            priors = []
+            for mod in self.pymc3_model:
+                with mod:
+                    priors.append(sample_prior_predictive(ndraws))
+            return priors
 
-    def sample_posterior(self, trace, ndraws=3):
+    def sample_posterior(self, ndraws=3):
         """
         Draw samples from the posterior distribution.
         :parameter n
         Number of posteriors to sample
         """
-        with self.model:
-            return sample_posterior_predictive(trace, ndraws)
+        if not hasattr(self, 'trace'):
+            print("Sampling has not been run yet! Running now with defaults...")
+            self.sample()
+
+        try:
```

Comment:
It might be good to be a little more explicit in this `try/except` statement. Any weird error that pops up will run the `except` commands, whereas I think you're mainly using this to determine "is `self.pymc3_model` a list or a PyMC3 model?" I think you could either add the one exception you'd be expecting to the `except`, or just change this to an `if` statement. (You can tell I've been burned before by bugs getting lost in open-ended `try` statements...)

Reply:
Done!